### PR TITLE
Fix tekton_pipelines_controller_taskrun_count recount bug

### LINKED ISSUE #4454: taskrun counter metrics grow up without failed taskrun
# Expected Behavior
tekton_pipelines_controller_taskrun_count{status="failed"} grow up only if taskrun failed and only increase by 1

# Actual Behavior
I have alerts about taskrun failed without failed taskrun and then check counter in prometheus and counter is growing up.

# Steps to Reproduce the Problem

Install tekton with helm and setup prometheus.

# Additional Info

- Kubernetes version:

  ```
Client Version: version.Info{Major:"1", Minor:"22", GitVersion:"v1.22.4", GitCommit:"b695d79d4f967c403a96986f1750a35eb75e75f1", GitTreeState:"clean", BuildDate:"2021-11-17T15:48:33Z", GoVersion:"go1.16.10", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"20+", GitVersion:"v1.20.10-gke.1600", GitCommit:"ef8e9f64449d73f9824ff5838cea80e21ec6c127", GitTreeState:"clean", BuildDate:"2021-09-06T09:24:20Z", GoVersion:"go1.15.15b5", Compiler:"gc", Platform:"linux/amd64"}
WARNING: version difference between client (1.22) and server (1.20) exceeds the supported minor version skew of +/-1
  ```

- Tekton Pipeline version:

  **v0.27.1% **

One more fing. For one minutes counter jumped too mach
1750 @1641551193.666
1764 @1641551223.673
1778 @1641551253.666

and for one taskfail i have grow counter with two
and Sorry for my english

### PR DESCRIPTION
**Pull request by khrm**
Added before and after condition check to avoid taskrun metrics recount bug.
Fixes #4454

# Changes

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [ ] [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [x] [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Release notes block below has been filled in or deleted (only if no user facing changes)

# Release Notes

```release-note
Fix `tekton_pipelines_controller_taskrun_count recount bug` recount bug which causes an increase in the number of taskrun without any new taskrun execution.
```

### CODE REVIEW

File: pkg/taskrunmetrics/metrics_test.go [hunk 1]
```diff
@@ -59,7 +59,7 @@ func getConfigContext() context.Context {
 func TestUninitializedMetrics(t *testing.T) {
 	metrics := Recorder{}
 
-	if err := metrics.DurationAndCount(&v1beta1.TaskRun{}); err == nil {
+	if err := metrics.DurationAndCount(&v1beta1.TaskRun{}, nil); err == nil {
```
Comment (pritidesai) [inline pkg/taskrunmetrics/metrics_test.go:62]: thank you @khrm for adding this, is it possible to add a separate test to validate metrics count when `before` condition is not nil?

Comment (khrm) [reply pkg/taskrunmetrics/metrics_test.go:62]: I added tests with `before` condition.

Comment (khrm) [reply pkg/taskrunmetrics/metrics_test.go:62]: I added `before` condition tests.

File: pkg/taskrunmetrics/metrics_test.go [hunk 2]
```diff
@@ -286,7 +363,7 @@ func TestRecordTaskRunDurationCount(t *testing.T) {
 				t.Fatalf("NewRecorder: %v", err)
 			}
 
-			if err := metrics.DurationAndCount(c.taskRun); err != nil {
+			if err := metrics.DurationAndCount(c.taskRun, nil); err != nil {
```
Comment (pritidesai) [inline pkg/taskrunmetrics/metrics_test.go:366]: this should have the `beforeCondition` instead of `nil` 🙃 

```
if err := metrics.DurationAndCount(c.taskRun, c.beforeCondition); err != nil {
```

File: pkg/taskrunmetrics/metrics_test.go [hunk 3]
```diff
@@ -164,6 +165,79 @@ func TestRecordTaskRunDurationCount(t *testing.T) {
...
+			"task":      "task-1",
+			"taskrun":   "taskrun-1",
+			"namespace": "ns",
+			"status":    "success",
+		},
+		expectedCountTags: map[string]string{
+			"status": "success",
+		},
+		expectedDuration: 60,
+		expectedCount:    1,
```
Comment (pritidesai) [inline pkg/taskrunmetrics/metrics_test.go:236]: should this be `0` since there is no change in `before` and `after` conditions?

File: pkg/taskrunmetrics/metrics_test.go (hunk 2 above)
Comment (khrm) [reply pkg/taskrunmetrics/metrics_test.go]: Yes, that's why test was passing. I fixed this.

File: pkg/taskrunmetrics/metrics_test.go (hunk 3 above)
Comment (khrm) [reply pkg/taskrunmetrics/metrics_test.go]: Yes, I fixed this.

**Review by pritidesai — APPROVED** [review]
thank you @khrm 👍
